File: modules/recipely-assistant-kit/android/src/main/java/net/recipely/assistantkit/RecipelyAssistantKitModule.kt
package net.recipely.assistantkit

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONArray
import org.json.JSONObject

/**
 * The JavaScript face of the shared store.
 *
 * Mirrors `RecipelyAssistantKitModule.swift` method for method, so a call site
 * never asks which platform it is on. This module never performs an action: it
 * hands invocations up to `AssistantActionRegistry`, the one thing that knows
 * how to run a word, and takes the app's recipe catalogue down.
 */
class RecipelyAssistantKitModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("RecipelyAssistantKit")

    Events(INVOCATION_EVENT)

    Property("isAvailable") { true }

    AsyncFunction("getPendingInvocationsAsync") {
      toMaps(RecipelyAssistantStore.pendingInvocations(context))
    }

    AsyncFunction("removePendingInvocationAsync") { invocationId: String ->
      RecipelyAssistantStore.removeInvocation(context, invocationId)
    }

    AsyncFunction("clearPendingInvocationsAsync") {
      RecipelyAssistantStore.clearInvocations(context)
    }

    AsyncFunction("setEntityCatalogAsync") { kind: String, entries: List<Map<String, Any?>> ->
      RecipelyAssistantStore.setEntities(context, kind, toJsonArray(entries))
    }

    AsyncFunction("refreshShortcutsAsync") {
      RecipelyShortcutPublisher.publish(context)
    }

    AsyncFunction("setCredentialsAsync") { credentials: Map<String, Any?> ->
      RecipelyAssistantStore.setCredentials(
        context,
        credentials["token"] as? String,
        credentials["languageCode"] as? String ?: "en",
      )
    }
  }

  private val context
    get() = requireNotNull(appContext.reactContext) { "React context is unavailable" }

  private fun toMaps(array: JSONArray): List<Map<String, Any?>> =
    (0 until array.length()).mapNotNull { index ->
      array.optJSONObject(index)?.let { entry ->
        entry.keys().asSequence().associateWith { key -> entry.get(key) }
      }
    }

  private fun toJsonArray(entries: List<Map<String, Any?>>): JSONArray =
    JSONArray().apply { entries.forEach { put(JSONObject(it)) } }

  private companion object {
    const val INVOCATION_EVENT = "onInvocation"
  }
}

File: modules/recipely-assistant-kit/ios/RecipelyAssistantKitModule.swift
import ExpoModulesCore

/// The JavaScript face of the shared container.
///
/// - Note: **This module never performs an action.** It hands invocations up to
///   `AssistantActionRegistry`, which is the one thing that knows how to run a
///   word, and takes the app's recipe catalogue down. Putting dispatch here
///   would have meant a second implementation of the vocabulary, in a language
///   with no access to the screens the handlers drive.
/// - Note: **`onInvocation` fires only for a running app.** The cold-launch case
///   is the queue, not the event: by the time a listener could be attached the
///   intent has long since run.
public final class RecipelyAssistantKitModule: Module {
  static let invocationEvent = "onInvocation"
  /// The catalogue kind the app publishes recipes under; `OsAssistantBridge` spells the same word.
  static let recipeEntityKind = "recipe"

  public func definition() -> ModuleDefinition {
    Name("RecipelyAssistantKit")

    Events(Self.invocationEvent)

    Property("isAvailable") {
      RecipelyAssistantStore.appGroupIdentifier != nil
    }

    AsyncFunction("getPendingInvocationsAsync") { () -> [[String: Any]] in
      RecipelyAssistantStore.pendingInvocations()
    }

    AsyncFunction("removePendingInvocationAsync") { (invocationId: String) in
      RecipelyAssistantStore.removeInvocation(invocationId: invocationId)
    }

    AsyncFunction("clearPendingInvocationsAsync") {
      RecipelyAssistantStore.clearInvocations()
    }

    AsyncFunction("setEntityCatalogAsync") { (kind: String, entries: [[String: Any]]) in
      RecipelyAssistantStore.setEntities(kind: kind, entries: entries)
      // The same list, in the one place iOS looks when someone searches their
      // phone rather than naming the app. Apple requires the app's name in
      // every Siri phrase, so Spotlight is the only surface where "baklava"
      // alone can reach us.
      if kind == RecipelyAssistantKitModule.recipeEntityKind, #available(iOS 15.1, *) {
        RecipelySpotlightIndex.publish(entries)
      }
    }

    // iOS publishes its entity catalogue through the shared container, which
    // `setEntityCatalogAsync` has already written; the Shortcuts phrases are
    // static. Android is where this does real work.
    AsyncFunction("refreshShortcutsAsync") {
      // Intentionally empty on iOS.
    }

    AsyncFunction("setCredentialsAsync") { (credentials: [String: Any]) in
      RecipelyAssistantStore.setCredentials(
        token: credentials["token"] as? String,
        languageCode: credentials["languageCode"] as? String ?? "en"
      )
    }
  }
}
